File: 2421-number-of-good-paths/2421-number-of-good-paths.py
class union_find:
    def __init__(self, n):
        self.parent = [x for x in range(n)]
        self.size = [1] * n
    def union(self, i, j):
        parentI, parentJ = self.find(i), self.find(j)
        if parentI == parentJ:
            return
        if self.size[parentI] > self.size[parentJ]:
            self.size[parentI] += self.size[parentJ]
            self.parent[parentJ] = parentI
        else:
            self.size[parentJ] += self.size[parentI]
            self.parent[parentI] = parentJ            
    def find(self, i):
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]
class Solution:
    def numberOfGoodPaths(self, vals: List[int], edges: List[List[int]]) -> int:
        neighbors = defaultdict(list)
        n = len(vals)
        for x, y in edges:
            neighbors[x].append(y)
            neighbors[y].append(x)
        union_set = union_find(n)
        ans = 0
        valToIdx = defaultdict(list)
        #start from 0
        for i, val in enumerate(vals):
            valToIdx[val].append(i)
        
        for val in sorted(valToIdx.keys()):
            for idx in valToIdx[val]:
                for nxt in neighbors[idx]:
                    if vals[nxt] <= vals[idx]:
                        union_set.union(nxt, idx)
            countOfConnected = defaultdict(int)
            for idx in valToIdx[val]:
                countOfConnected[union_set.find(idx)] += 1
            
            for x in countOfConnected.values():
                ans += x + x*(x-1) // 2
            
        return ans
            
            
            
            

File: 2421-number-of-good-paths/2421-number-of-good-paths.java
class Solution {
    int[] size;
    int[] parent;
    public int find(int node) {
        if (node == parent[node])
            return node;
        parent[node] = find(parent[node]);
        return parent[node];
    }
    public int numberOfGoodPaths(int[] vals, int[][] edges) {
        int n = vals.length;
        int ans = n;
        Map<Integer, List<Integer>> graph = new HashMap<>();
        //start from the minimum vals
        int[][] valIdx = new int[n][2];
        
        for (int idx = 0; idx < n; idx++) {
            valIdx[idx][0] = vals[idx];
            valIdx[idx][1] = idx;
        }
        Arrays.sort(valIdx, (a, b) -> a[0] - b[0]);
        buildGraph(edges, n, graph);
        //size is num of maxVal in the union
        size = new int[n];
        parent = new int[n];
        
        for (int idx = 0; idx < n; idx++) {
            parent[idx] = idx;
            size[idx] = 1;
        }
        
        for (int idx = 0; idx < n; idx++) {
            int curVal = valIdx[idx][0];
            int realIdx = valIdx[idx][1];
            int curParent = find(realIdx);
            
            for (int nxt : graph.get(realIdx)) {
                int nxtVal = vals[nxt];
                int nxtParent = find(nxt);
                
                if (nxtParent == curParent || nxtVal > curVal)
                    continue;
                if (vals[nxtParent] == curVal) {
                    ans += size[nxtParent] * size[curParent];
                    size[curParent] += size[nxtParent];
                }
                parent[nxtParent] = curParent;

            }
        }
        
        return ans;

    }
    
    public void buildGraph(int[][] edges, int n, Map<Integer, List<Integer>> graph) {
        for (int idx = 0; idx < n; idx++) {
            graph.put(idx, new LinkedList<>());
        }
        
        for (int[] edge : edges) {
            int start = edge[0], end = edge[1];
            graph.get(start).add(end);
            graph.get(end).add(start);
        }
    }

}